lib/Dialect/Arc/Transforms/MergeIfs.cpp

```diff
@@ -0,0 +1,365 @@
+    return write.getMemory();
+  return {};
+}
+
+/// Return the state/memory value being read by an op.
+static Value getPointerReadByOp(Operation *op) {
+  if (auto read = dyn_cast<StateReadOp>(op))
+    return read.getState();
+  if (auto read = dyn_cast<MemoryReadOp>(op))
+    return read.getMemory();
+  return {};
+}
+
+namespace {
+/// An integer indicating the position of an operation in its parent block. The
+/// first field is the initial order/position assigned. The second field is used
+/// to order ops that were moved to the same location, which makes them have
+/// same first field.
+using OpOrder = std::pair<unsigned, unsigned>;
+
+/// A helper that tracks an op and its order, and allows for convenient
+/// substitution with another op that has a higher/lower order.
+struct OpAndOrder {
+  Operation *op = nullptr;
+  OpOrder order = {0, 0};
+
+  explicit operator bool() const { return op; }
+
+  /// Assign `other` if its order is lower than this op, or this op is null.
+  void minimize(const OpAndOrder &other) {
+    if (!op || (other.op && other.order < order))
+      *this = other;
+  }
+
+  /// Assign `other` if its order is higher than this op, or this op is null.
+  void maximize(const OpAndOrder &other) {
+    if (!op || (other.op && other.order > order))
+      *this = other;
+  }
+};
+} // namespace
+
+/// Sink operations as close to their users as possible.
+void Sinker::sinkOps() {
+  // A numeric position assigned to ops as we encounter them. Ops at the end of
+  // the block get the lowest order number, ops at the beginning the highest.
+  DenseMap<Operation *, OpOrder> opOrder;
+  // A lookup table that indicates where ops should be inserted. This is used to
+  // maintain the original op order if multiple ops pile up before the same
+  // other op that blocks their move.
+  DenseMap<Operation *, Operation *> insertionPoints;
+  // The write ops to each state/memory pointer we've seen so far. ("Next"
+  // because we run from the end to the beginning of the block.)
+  DenseMap<Value, Operation *> nextWrite;
+  // The most recent op that has an unknown (non-read/write) side-effect.
+  Operation *nextSideEffect = nullptr;
+
+  for (auto &op : llvm::make_early_inc_range(llvm::reverse(rootBlock))) {
+    // Assign an order to this op.
+    auto order = OpOrder{opOrder.size() + 1, 0};
+    opOrder[&op] = order;
+
+    // Analyze the side effects in the op.
+    op.walk([&](Operation *subOp) {
+      if (auto ptr = getPointerWrittenByOp(subOp))
+        nextWrite[ptr] = &op;
+      else if (!isa<StateReadOp, MemoryReadOp>(subOp) &&
+               !subOp->hasTrait<OpTrait::HasRecursiveMemoryEffects>() &&
+               !mlir::isMemoryEffectFree(subOp))
+        nextSideEffect = &op;
+    });
+
+    // Determine how much the op can be moved.
+    OpAndOrder moveLimit;
+    if (auto ptr = getPointerReadByOp(&op)) {
+      // Don't move across writes to the same state/memory.
+      if (auto *write = nextWrite.lookup(ptr))
```

Comment:
This assumes no aliasing can happen. It's probably fine in the current arcilator pipeline, but we should maybe make that more explicit in the pass documentation.

Reply:
Yeah that's a great point. Maybe we should make that a feature of the dialect itself somewhere? As in, document that states and memories are assumed to never alias. Not sure how well that will work with a more struct-like memory layout though, where we'll need something like `arc.state_ptr %storage, @DataLayout::someField` instead of `arc.alloc_state`. Those storage-to-state-pointer ops probably will get duplicated and it'll be harder to know about the aliasing. Might need to hook into MLIR's alias analysis at that point.

I'll add something to the pass' description for now.

Reply:
Yes, good point. That would be a nice property for the dialect, but not sure if it would restrict us too much in the struct case you mentioned. I guess we'll see.

Your storage-to-state-pointer duplication is more like what I thought of here. It could also be an `scf.if` returning a state or a call or any other op that could forward a state value in some way. MLIR's alias analysis should make this easy to support properly once we need it, though.